source/common/http/mixed_conn_pool.cc

```diff
@@ -0,0 +1,74 @@
+#include "common/http/mixed_conn_pool.h"
+
+#include "common/http/codec_client.h"
+#include "common/http/http1/conn_pool.h"
+#include "common/http/http2/conn_pool.h"
+#include "common/http/utility.h"
+#include "common/tcp/conn_pool.h"
+
+namespace Envoy {
+namespace Http {
+
+Envoy::ConnectionPool::ActiveClientPtr HttpConnPoolImplMixed::instantiateActiveClient() {
+  return std::make_unique<Tcp::ActiveTcpClient>(*this,
+                                                Envoy::ConnectionPool::ConnPoolImplBase::host(), 1);
+}
+
+CodecClientPtr
+HttpConnPoolImplMixed::createCodecClient(Upstream::Host::CreateConnectionData& data) {
+  auto protocol =
+      protocol_ == Protocol::Http11 ? CodecClient::Type::HTTP1 : CodecClient::Type::HTTP2;
+  CodecClientPtr codec{new CodecClientProd(protocol, std::move(data.connection_),
+                                           data.host_description_, dispatcher_, random_generator_)};
+  return codec;
+}
+
+void HttpConnPoolImplMixed::onConnected(Envoy::ConnectionPool::ActiveClient& client) {
+  // When we upgrade from a TCP client to non-TCP we get a spurious onConnected
+  // from the new client. Ignore that.
```

Comment:
This seems broken. TODO somewhere to not raise onConnected() multiple times?

Reply:
What happens here is that the client is subscribed to network callbacks. When under the call stack of the network raising the connection event, we detach the TCP client from the network callbacks, and attach the HTTP client.

I don't think it's safe to delay subscribing the HTTP client to the network callbacks until not under the stack of raising the connected event.  We could avoid the detach and reattach by having a shim event handler in the connection class which originally pass onEvent to the TCP client, and is std::moved over to pass onEvent to the HTTP client.  But the problem with that is that for TCP, the connection pool client subscribes directly to the network's callbacks.  For HTTP, the connection pool client asks the codec to add it to network callbacks, and doesn't do so to the connection directly.  Today that's the same code, but if we decide to do custom event work in the codec client, we could end up introducing some really weird bugs.

Long story short I actually think this is cleaner than any other option.  If I can sell you on that I'd be happy to add inline comments on why it's safer than other options.  Otherwise I'd prefer nailing down something which isn't uglier before I add a TODO :-)

Reply:
Sorry are you saying that because we are under the same call stack the event gets added and then it gets iterated to and called? I'm surprised that actually works but I forget what data structure is used. I believe you that this is the cleanest way so yeah more comments would be good.